### PR TITLE
Exclude certain languages from repo stats

### PR DESCRIPTION
**Pull request by CirnoT**
For one reason or another, Linguist sometimes decides that it is better to not mark file as generated, so that its diff is not suppressed on GitHub. 

One such example is `yarn.lock`, often present in Node repositories using Yarn as package manager. Due to it's nature it was decided to not include it as generated file as it is easy to inspect its diff for security purposes (lockfile poison).

---

However, despite that, GitHub still does not include such files in their language bar, unless they are the only file present in repository.

Further tests revealed that in fact, GitHub explicitly ignores certain languages unless they have 100% presence in repository. There is no publicly available list of such exclusions, however prominent examples are JSON and YAML - you can verify this by trying to search GitHub repositories by `language:`.

This PR aims to add support for similar logic in Gitea. The list of languages present right now might not be full, the only real way to find would be to test every language GitHub knows about.

### CODE REVIEW

**Comment by silverwind** [comment]
I wonder if this should be better done upstream in [go-enry](https://github.com/go-enry/go-enry).

**Comment by CirnoT** [comment]
`go-enry` aims to replicate linguist behavior, and linguist decided that `yarn.lock` should not be marked as generated back in 2019. 

Doing this in upstream would mean significant changes between linguist and go-enry, and I believe that is not the aim of the project.

Additionally, GitHub also performs same exclusion outside of linguist scope (otherwise we'd be seeing lot of YAML repos instead of JavaScript ones)

---

ref https://github.com/github/linguist/issues/4348 https://github.com/github/linguist/pull/4445 https://github.com/github/linguist/pull/4459 https://github.com/github/linguist/commit/3bc818554f16d52a3867b37c0b36e674202ca999

**Comment by CirnoT** [comment]
Frankly I find reasoning given by linguist insufficient in consideration that every other lockfile in existence is excluded EXCEPT for `yarn.lock` because... someone didn't want diffs for it collapses and linguist maintainers agreed.

Either way, I think this PR is still valid as it aims to improve code statistics detection by removing unnecessary noise, bringing it closer to GitLab and GitHub's implementations.

**Comment by silverwind** [comment]
I'd definitely agree that `yarn.lock` should be considered generated and it just seems silly to not do it for `yarn.lock` but do so for `package-lock.json`. 

The argument that yaml is easier to diff than json is also silly, I can easily visually diff both for small changes and for bigger ones I skip over both of them.

**Comment by silverwind** [comment]
Maybe add a comment describing that the reason for the existance of this list is to overrule linguist/enry's behaviour that is considered suboptimal.

**Comment by silverwind** [comment]
What I worry thought is that let's say I have a repo for a Ansible role. These are almost purely YAML and those files are basically equivalent to code. If I introduce a bash script into said repo, it would no longer count any YAML at all?

Maybe some sort of filename-based approach would be better and for that we can mostly lean on enry/linguist, right?

**Comment by CirnoT** [comment]
That isn't strictly true; the existence of the list can be owned to linguist poor choice because that's what personally made me write the code, but in general the purpose of the list for now and for future is to not only overrule but also remove noise.

I would rather not implement any file-based approach lest we re-do linguist/enry on our own. I advise you to push example repository to GitHub and see how they see it. My bet is that they will ignore YAML just like my implementation.

**Comment by CirnoT** [comment]
I found good example here: https://github.com/ansible/ansible-examples

**Comment by CirnoT** [comment]
GitLab outright refuses to even acknowledge YAML even if it's lone in repository: https://gitlab.com/maksold/ansible-role-certbot

**Comment by silverwind** [comment]
So you're saying this exclusion is not part of linguist but some hackery that GitHub does after the files were processed by linuist?

I wonder if one can overrule this via gitattributes `*.yaml -linguist-vendored -linguist-generated`.

Maybe instead of a plain exclusion list, we want to specify the files as generated so enry will see them as such, it'd be more compatible and user-overridable.

**Comment by CirnoT** [comment]
Yes you can overrule it for linguist, it will make GitHub show full diff as it won't treat these files as generated/vendored, but it doesn't seem to affect language statistics.

GitLab doesn't use linguist at all and I'm not sure you can even override it there.

**Comment by CirnoT** [comment]
These ansible files aren't considered generated or vendored by linguist in the first place, so `.gitattributes` change has no meaning really. It's just that GitHub explicitly ignores specific languages from language bar/statistics, and so does GitLab. It helps keep the noise down I suppose.

**Comment by CirnoT** [comment]
Generally GitHub uses linguist for few different things. One of them is deciding if diff should be collapsed, another is language bar statistics. For language bar they apply additional logic to clean out certain languages from showing in almost every repo.

**Comment by silverwind** [comment]
Ok, I guess the notion behind this that these files would never represent source code. Certainly not true for YAML or SVG but those are kind of grey areas.

BTW, yarn lockfile v1 is not strictly YAML format, only the v2 version is. Still most mechanisms seem to pick it up as YAML for some reason.

**Comment by CirnoT** [comment]
Actually linguist explicitly marks `yarn.lock` filename as YAML, for similar reason the generated exclusion was disabled - syntax coloring for diffs.

File: modules/git/repo_language_stats.go [hunk 1]
```diff
@@ -20,6 +20,16 @@ import (
 
 const fileSizeLimit int64 = 16 * 1024 * 1024
 
+// specialLanguages defines list of languages that are excluded from calculation
+// unless they are the only language present in repository
```
Comment (silverwind) [inline modules/git/repo_language_stats.go]:
```suggestion
// specialLanguages defines list of languages that are excluded from the calculation
// unless they are the only language present in repository. Only languages which under
// normal circumstances are not considered to be code should be listed here.
```

**Comment by jolheiser** [comment]
Would this be better served as a configurable option?

`.gitea/stats` file in the repo or something like that?

This would allow us to close #10266 (alternatively you could add TOML to that list, but I feel like a per-repo config file might be more future-proof)

**Comment by CirnoT** [comment]
I'm generally opposed to idea of configuring language statistics per-repo in yet-another-dot-file(r)

**Comment by zeripath** [comment]
although I'm generally opposed to requiring yet another dotfile - as I think we should generally just get things right first time - I think having an option to override behaviour is always a good idea.

That being said - that can be done as a different PR.

**Comment by lafriks** [comment]
No need for additional dot file. Github does language stats fine-tuning using `.gitattributes` file but imho that is out of scope for this PR

**Comment by CirnoT** [comment]
I won't oppose any PR that adds such ability via `.gitattributes` if it's opt-out.

**Review by silverwind — APPROVED** [review]

**Comment by CirnoT** [comment]
Deprecated by https://github.com/go-gitea/gitea/pull/11681

**Pull request closed without merging**